### PR TITLE
add push support

### PR DESCRIPTION
**Pull request by allcloud-jonathan**
solves #8

### CODE REVIEW

**Comment by allcloud-jonathan** [comment]
@johananl please take a look

File: onelogin/get.go [hunk 1]
```diff
@@ -107,30 +109,80 @@ func Get(app, provider string) (*awsprovider.Credentials, error) {
...
-	fmt.Scanln(&otp)
+	var samlAssertion string
+
+	var pushSupported = false
+
+	if deviceType == "OneLogin Protect" {
+		// push is supported by the selected OTP device
+		// if other are supporting push they should be added here.
+		pushSupported = true
+		s.Start()
```
Comment (johananl) [inline onelogin/get.go]: There are currently visual problems in the terminal because the spinner is started and stopped at the wrong places.
Following is the correct behavior:

```go
s.Start()
rMfa, err := c.VerifyFactor(token, &pMfa)
if err != nil {
    return nil, err
}
s.Stop()

pMfa.DoNotNotify = true

timeout := 30
interval := 4
fmt.Println(rMfa.Status.Message)
s.Start()
for rMfa.Status.Type == "pending" && timeout > 0 {
    time.Sleep(time.Duration(interval) * time.Second)
    rMfa, err = c.VerifyFactor(token, &pMfa)
    if err != nil {
        s.Stop()
        return nil, err
    }

    timeout -= interval
}
s.Stop()

if rMfa.Status.Type == "pending" {
    fmt.Println("Push timed out, please enter OTP token.")
    pushSupported = false
} else {
    samlAssertion = rMfa.Data
}
```

File: onelogin/get.go [hunk 2]
```diff
@@ -107,30 +109,80 @@ func Get(app, provider string) (*awsprovider.Credentials, error) {
...
 		deviceID = fmt.Sprintf("%v", devices[0].DeviceId)
+		deviceType = devices[0].DeviceType
 	}
 
-	fmt.Print("Please enter the OTP from your MFA device: ")
-	var otp string
-	fmt.Scanln(&otp)
+	var samlAssertion string
+
+	var pushSupported = false
```
Comment (johananl) [inline onelogin/get.go]: I would rename this var to `pushOK` or `pushSuccessful`.
We actually check if push is **supported** by checking the device type. This var only tells us whether the user was able to accept the push on their phone.

File: onelogin/get.go [hunk 3]
```diff
@@ -107,30 +109,80 @@ func Get(app, provider string) (*awsprovider.Credentials, error) {
...
-	}
+		rMfa, err := c.VerifyFactor(token, &pMfa)
+		if err != nil {
+			return nil, err
+		}
+		pMfa.DoNotNotify = true
+		timeout := 30
+		interval := 4
+		fmt.Printf(" %v\r", rMfa.Status.Message)
+		for rMfa.Status.Type == "pending" && timeout > 0 {
```
Comment (johananl) [inline onelogin/get.go]: I would have liked to fail fast here in case the user presses "reject" on the app, however looks like the status remains "pending" both when the user doesn't do anything and when they press "reject"...
Any idea if we can get a "reject" indication?

File: onelogin/get.go [hunk 4]
```diff
@@ -107,30 +109,80 @@ func Get(app, provider string) (*awsprovider.Credentials, error) {
...
+			if err != nil {
+				return nil, err
+			}
+
+			fmt.Printf(" %v\r", rMfa.Status.Message)
+			timeout -= interval
+		}
+		if rMfa.Status.Type == "pending" {
+			s.Stop()
+			fmt.Println("Push timed out, please enter OTP token.")
```
Comment (johananl) [inline onelogin/get.go]: This message is a bit redundant with the next one. I would rephrase to:

`MFA verification failed - falling back to manual OTP input`

File: onelogin/get.go [hunk 5]
```diff
@@ -107,30 +109,80 @@ func Get(app, provider string) (*awsprovider.Credentials, error) {
...
-	fmt.Print("Please enter the OTP from your MFA device: ")
-	var otp string
-	fmt.Scanln(&otp)
+	var samlAssertion string
+
+	var pushSupported = false
+
+	if deviceType == "OneLogin Protect" {
+		// push is supported by the selected OTP device
+		// if other are supporting push they should be added here.
```
[the remark this answers is not in the document]
Comment (allcloud-jonathan) [reply onelogin/get.go]: We'll I wasn't sure about OneLogin SMS. As it's not enabled in our corp, I wouldn't know.

File: onelogin/get.go (hunk 1 above)
Comment (allcloud-jonathan) [reply onelogin/get.go]: It was actually my intention to have the spinner running while outputting the status message. That's why I added a space at the beginning and `\r` at end of the status output. I mean I've never seen another message than the one displayed, but the intention was to display a possibly changing message and I didn't want to have multiple lines of the same message.

File: onelogin/get.go (hunk 3 above)
Comment (allcloud-jonathan) [reply onelogin/get.go]: nope. if the status changes we're leaving the loop. Maybe a future API update. ;-)

File: onelogin/get.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (johananl) [reply onelogin/get.go]: Sounds from the docs that only Protect currently supports push. SMS has a different mechanism and since we haven't implemented it in the app yet, I guess we can simply ignore this case.
In any case, I think this comment should be edited.

File: onelogin/get.go (hunk 4 above)
Comment (allcloud-jonathan) [reply onelogin/get.go]: I slightly reworded that to `MFA verification timed out - falling back to manual OTP input`

File: onelogin/get.go (hunk 1 above)
Comment (johananl) [reply onelogin/get.go]: Yeah, I figured this is what you wanted.
The problem is - it creates visual problems as it is now. I prefer to have something a bit less fancy but stable.
And in any case, the spinner should start right before the long-running piece of code and stop as soon as possible.

File: onelogin/get.go (hunk 3 above)
Comment (johananl) [reply onelogin/get.go]: The thing is - pressing "reject" doesn't make us leave the loop, we only leave when it times out...

File: onelogin/get.go (hunk 4 above)
Comment (johananl) [reply onelogin/get.go]: But that's the thing - this is also printed when the user hits "reject", so this is why I wrote "failed" and not "timed out"...

File: onelogin/get.go [hunk 6]
```diff
@@ -107,36 +121,85 @@ func Get(app, provider string) (*awsprovider.Credentials, error) {
...
+		}
+
+		rMfa, err = c.VerifyFactor(token, &pMfa)
+		if err != nil {
+			return nil, err
+		}
+
+		pMfa.DoNotNotify = true
+
+		s.Stop()
```
Comment (johananl) [inline onelogin/get.go]: This one should be right *after* we called `c.VerifyFactor`.

File: onelogin/get.go [hunk 7]
```diff
@@ -107,36 +121,85 @@ func Get(app, provider string) (*awsprovider.Credentials, error) {
...
+		fmt.Println(rMfa.Status.Message)
+		s.Start()
```
Comment (johananl) [inline onelogin/get.go]: This should be right before we start sleeping in a loop. At the moment there is a var assignment in the middle...

File: onelogin/get.go [hunk 8]
```diff
@@ -107,36 +121,85 @@ func Get(app, provider string) (*awsprovider.Credentials, error) {
...
+
+		timeout := MFAPushTimeout
+		for rMfa.Status.Type == "pending" && timeout > 0 {
+			time.Sleep(time.Duration(MFAInterval) * time.Second)
+			rMfa, err = c.VerifyFactor(token, &pMfa)
+			if err != nil {
```
Comment (johananl) [inline onelogin/get.go]: We should `s.Stop()` here before returning.

File: onelogin/get.go [hunk 9]
```diff
@@ -107,36 +121,85 @@ func Get(app, provider string) (*awsprovider.Credentials, error) {
...
+				return nil, err
+			}
+
+			timeout -= MFAInterval
+		}
+		if rMfa.Status.Type == "pending" {
+			s.Stop()
```
Comment (johananl) [inline onelogin/get.go]: This should be before the `if`. We want to stop anyway.

**Comment by johananl** [comment]
LGTM.
